packages/syft/src/syft/core/node/common/action/smpc_action_message.py

```diff
@@ -0,0 +1,291 @@
+import numpy as np
+
+# relative
+# syft relative
+from ..... import serialize
+from .....proto.core.node.common.action.smpc_action_message_pb2 import (
+    SMPCActionMessage as SMPCActionMessage_PB,
+)
+from ....common.message import ImmediateSyftMessageWithoutReply
+from ....common.serde.deserialize import _deserialize
+from ....common.serde.serializable import bind_protobuf
+from ....common.uid import UID
+from ....io.address import Address
+from ....tensor.smpc.share_tensor import ShareTensor
+
+# How many intermediary ids we generate in each smpc function
+MAP_FUNC_TO_NR_GENERATOR_INVOKES = {"__add__": 0, "__mul__": 0, "__sub__": 0}
+
+
+@bind_protobuf
+class SMPCActionMessage(ImmediateSyftMessageWithoutReply):
+    def __init__(
+        self,
+        name_action: str,
+        self_id: UID,
+        args_id: List[UID],
+        kwargs_id: Dict[str, UID],
+        result_id: UID,
+        address: Address,
+        ranks_to_run_action: Optional[List[int]] = None,
+        msg_id: Optional[UID] = None,
+    ) -> None:
+        self.name_action = name_action
+        self.self_id = self_id
+        self.args_id = args_id
+        self.kwargs_id = kwargs_id
+        self.id_at_location = result_id
+        self.ranks_to_run_action = ranks_to_run_action
+        self.address = address
+        self.msg_id = msg_id
+        super().__init__(address=address, msg_id=msg_id)
+
+    @staticmethod
+    def filter_actions_after_rank(
+        rank: int, actions: List["SMPCActionMessage"]
+    ) -> List["SMPCActionMessage"]:
+        """
+        Filter the actions depending on the rank of each party
+
+        Arguments:
+            rank (int): the rank of the party
+            actions (List[SMPCActionMessage]):
+
+        """
+        res_actions = []
+        for action in actions:
+            if action.ranks_to_run_action is None:
+                raise ValueError(
+                    "Attribute ranks_to_run_action should not be None when filtering"
+                )
+
+            if action.ranks_to_run_action == [] or rank in action.ranks_to_run_action:
+                res_actions.append(action)
+        return res_actions
+
+    @staticmethod
+    def get_action_generator_from_op(
+        operation_str: str,
+    ) -> Callable[[UID, UID, int, Any], Any]:
+        """ "
+        Get the generator for the operation provided by the argument
+        Arguments:
+            operation_str (str): the name of the operation
+
+        """
+        return MAP_FUNC_TO_ACTION[operation_str]
+
+    def _object2proto(self) -> SMPCActionMessage_PB:
+        """Returns a protobuf serialization of self.
+
+        As a requirement of all objects which inherit from Serializable,
+        this method transforms the current object into the corresponding
+        Protobuf object so that it can be further serialized.
+
+        :return: returns a protobuf object
+        :rtype: RunClassMethodAction_PB
```

Comment:
Maybe we can do a quick find replace on the copied `RunClassMethodAction` string inside this file?

Reply:
I have done this.